### ISSUE TITLE
Moving Elide toward supporting Knowledge Graph data CRUD

### ISSUE THREAD

**Issue by QubitPi** [paion-data/elide#35]
Elide has been used in Yahoo for nearly 10 years. Is has been used to support lots of businesses at Yahoo successfully.

One of the key cutting-edge advantage of Elide is its extremely flexible architecture that not only support SQL databases, which Elide was initially designed for, but also it has successfully extended its capability to non-SQL world such as [text search](https://github.com/yahoo/elide/tree/master/elide-datastore/elide-datastore-search), the one very similar to Elasticsearch & [Apache Druid](https://druid.apache.org/) through its [analytic module](https://elide.io/pages/guide/v7/04-analytics.html)

As a startup that constantly focuses on knowledge graph application and technology innovation, we would like Elide to natively support the CRUD of a new type of data - [graph data](https://en.wikipedia.org/wiki/Graph_theory#Applications), this would bring at least 2 benefits

1. keep Paion on the cutting edge of exploring new technologies and opportunities
2. bring this Yahoo's classical software to Chinese tech market by the opportunity of knowledge graph tech

Ths initiative of supporting graph data by Elide will benefit company's most important product - [Nexus Graph](https://github.com/paion-data/nexusgraph)

Short-Term Goal - Storing Graph Data in SQL Database
----------------------------------------------------

The relationship between a graph node and link is definitely __one-to-many__, given that one node can have multiple outgoing/incident links.

Intuitively one might model node and link in the following way

```java
@Entity
@Table(name = "Node")
@Include(rootLevel = false, name = "node", description = "graph node", friendlyName = "node")
public class Node {

    ...

    /**
     * All incoming and outgoing links attached to this node.
     */
    @OneToMany(cascade = CascadeType.ALL)
    @GraphQLDescription("All incoming and outgoing links attached to this node.")
    public List<Link> links;
}
```

```java
@Entity
@Table(name = "Link")
@Include(rootLevel = false, name = "link", description = "graph link", friendlyName = "link")
public class Link {

    ...

    /**
     * The source node.
     */
    @ManyToOne
    @GraphQLDescription("The source node.")
    public Node source;

    /**
     * The target node.
     */
    @ManyToOne
    @GraphQLDescription("The target node.")
    public Node target;
}
```

There are at least 3 problems with this design

1. Semantically, `Node.links` is [mappedBy](https://stackoverflow.com/a/21068644) both `Link.source` and `Link.target` but `mappedBy` can only take one of them
2. By experiment, there is no way to create a knowledge graph in one GraphQL mutation.

   ```graphql
   mutation {
      node {
         id
      }
      link(id: <-- node.id) {
         id
      }
   }
   ```

   Creating a graph with nodes and links requires the node ID, upon created, to be passed to the link so that it can
   reference to create source and target node.
   
   In GraphQL, fields at each "level" of the request are executed and resolved in parallel. In this example, both `node`
   and `link` are fields of the same type so they will be resolved at the same time. That means each query essentially
   is not aware of the other or what the other resolves to.
   
   We could make nested query but nodes and links are intrinsically not nested, because a node with a link comes with
   another node for sure and two nodes are definitely on the same "level" which cannot be resolved with a deterministic
   order.
   
   We would have to handle this kind of scenario at client (Nexus Graph) side. In other words, create nodes first, parse
   the response for the node ID's and then make a second request.

3. An even more devastating issue is _it makes them no longer @OneToMany relationship_, because given that a link can connect up to 2 nodes, the owning-side (i.e. Link) of the foreign key table will have two rows with the same link ID, which is not possible. For example, suppose we have two nodes A, B with ID's 1 and 2, connected by 3 unidirectional links whose ID's are 11, 12, 13

   ```
                  10              
     ┌─────────────────────────┐  
     │                         │  
   ┌─▼─┐                     ┌─┴─┐
   │   │          11         │   │
   │ A ◄─────────────────────┤ B │
   │   │                     │   │
   └─▲─┘                     └─┬─┘
     │            12           │  
     └─────────────────────────┘  
   ```

   This would result in a join table like

   ```
   +---------+---------+
   | node_id | link_id |
   +---------+---------+
   |       1 |      10 |
   |       1 |      11 |
   |       1 |      12 |
   |       2 |      10 |
   |       2 |      11 |
   |       2 |      12 |
   +---------+---------+
   ```  
   
   __The table above presents a many-to-many relationship which breaks the one-to-many between node and links__
   
Given the difficulty of Elide in such business case, the short-term goal is to __design a JPA model of knowledge graph__ such that all CRUD operations against a knowledge graph can be down _in one GraphQL operations_

Mid-Term Goal - Storing Graph Data in Neo4J/Arango Graph Database
-----------------------------------------------------------------

Elide started as a JPA web service by exposing relational data store via a relational JPA data model. _Everything is relational initially_. It is important to realize that Elide, later, supported non-relational data CRUD in one of the two approaches

1. [Relational JPA data model backed by Non-relational data store](https://github.com/yahoo/elide/tree/master/elide-datastore/elide-datastore-search)

   The search data store is essentially [Apache Lucene](https://lucene.apache.org/), the store behind Elasticsearch, wrapped by Hibernate. Basically everything in Elide is still relational at this point

2. [Non-relational data model backed by relational data store](https://elide.io/pages/guide/v7/04-analytics.html)

   As Elide continues evolving, it started supporting CRUD operation against non-SQL database via [analytic query](https://elide.io/pages/guide/v7/04-analytics.html), which is commonly used in Business Intelligence applications like the one shown below:

   ![yavin](https://github.com/paion-data/elide/assets/16126939/5b7462db-54bf-43dd-9157-454fbb6e6f88)

   Basically, Elide achieves this through its concept of [semantic layer](https://elide.io/pages/guide/v7/04-analytics.html), which maps a relational view of data to arbitrary non-relational view for user

Working toward graph data, Paion Data will make Elide support a new and 3rd mechanism:

3. __[Non-relational data model](https://graph.build/resources/ontology) backed by non-relational data store ([Neo4J](https://neo4j.com/product/neo4j-graph-database/)/[ArangoDB](https://arangodb.com/))__

Long-Term Goal - Unified API Layer Aggregating Heterogeneous Databases
----------------------------------------------------------------------

In the long-run, Nexus Graph will need to deal with data from variety of sources, including AI-inferenced data, user generated knowledge graph data, and graph data from the world wide web. Each type of data will be stored in different types of databases. The key to make our product successfully evolve is to __unify the data API for efficient data management__. Elide will ultimately become the single data API aggregating arbitrarily heterogeneous data sources for this purpose

**Comment by Doom9527** [comment]
A possible solution is to use a linking table to store the `Link` entities, which includes two foreign key fields pointing to the `source` and `target` nodes respectively. This way, you can maintain a one-to-many relationship.
<img width="212" alt="1720279429072" src="https://github.com/paion-data/elide/assets/164460473/17f9d6b7-905a-48c3-9f85-78135095cf9b">

**Comment by QubitPi** [comment]
What would GraphQL query look like in the following case

1. Creating a new graph in one request
2. Updating a new graph in one request

**Comment by QubitPi** [comment]
> A possible solution is to use a linking table to store the `Link` entities, which includes two foreign key fields pointing to the `source` and `target` nodes respectively. This way, you can maintain a one-to-many relationship. <img alt="1720279429072" width="212" src="https://private-user-images.githubusercontent.com/164460473/346282903-17f9d6b7-905a-48c3-9f85-78135095cf9b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAyNzk4NzYsIm5iZiI6MTcyMDI3OTU3NiwicGF0aCI6Ii8xNjQ0NjA0NzMvMzQ2MjgyOTAzLTE3ZjlkNmI3LTkwNWEtNDhjMy05Zjg1LTc4MTM1MDk1Y2Y5Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcwNlQxNTI2MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNjVhYzVjMmI5ZmZjZWFiMjEwMTVlYzYxYzhkZDVjZGExMTQ4MmIxMzE1NGRlOTI5ZDg1YjQ0YTllZDczYTMyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.kMT9CEtuco-4mOvgWhlzBArT8tfO_DiB6eZfE6mVdhg">

This would not be considered as a "solution" because

1. It's incomplete
2. What's your definition of a "graph"?
3. Does this design allow atomicity of all possible graph requests mentioned above?

**Comment by QubitPi** [comment]
It also helps to write a draft of your data model with some code snippet. That helps your audience better understand your idea

**Comment by Doom9527** [comment]
The GraphQL query will like:
Creating a new graph in one request
```
mutation CreateGraph {
  createSource(input: {name: "Node A"}) {
    id
  }
  createTarget(input: {name: "Node B"}) {
    id
  }
  createLink(input: {sourceId: "nodeAId", targetId: "nodeBId"}) {
    id
    source {
      id
      name
    }
    target {
      id
      name
    }
  }
}
```
Updating a new graph in one request
```
mutation UpdateGraph {
  updateSource(id: "nodeAId", input: {name: "Updated Node A"}) {
    id
    name
  }
}
```
Fetching graph by ID
```
query GetGraphById($nodeId: ID!) {
  source(id: $nodeId) {
    id
    name
    links {
      id
      source {
        id
        name
      }
      target {
        id
        name
      }
    }
  }
}
```
Fetching all nodes of a graph without full table scan
```
query GetAllNodes {
  allSources {
    edges {
      node {
        id
        name
        links {
          edges {
            node {
              id
              source {
                id
                name
              }
              target {
                id
                name
              }
            }
          }
        }
      }
    }
  }
}
```

**Comment by QubitPi** [comment]
> The GraphQL query will like: Creating a new graph in one request
> 
> ```
> 
> ```
> 
> Updating a new graph in one request
> 
> ```
> mutation UpdateGraph {
>   updateSource(id: "nodeAId", input: {name: "Updated Node A"}) {
>     id
>     name
>   }
> }
> ```
> 
> Fetching graph by ID
> 
> ```
> query GetGraphById($nodeId: ID!) {
>   source(id: $nodeId) {
>     id
>     name
>     links {
>       id
>       source {
>         id
>         name
>       }
>       target {
>         id
>         name
>       }
>     }
>   }
> }
> ```
> 
> Fetching all nodes of a graph without full table scan
> 
> ```
> query GetAllNodes {
>   allSources {
>     edges {
>       node {
>         id
>         name
>         links {
>           edges {
>             node {
>               id
>               source {
>                 id
>                 name
>               }
>               target {
>                 id
>                 name
>               }
>             }
>           }
>         }
>       }
>     }
>   }
> }
> ```

in updating, what if we added 2 new nodes with 1 unidirectional link between then and at the same time updated the name of a 3rd node and yet also updated the label of another link?

**Comment by Doom9527** [comment]
This situation like 'fields at each "level" of the request are executed and resolved in parallel', I can not send it in one request, I had to think about designing a new JPA model or using some other solution.

**Comment by Doom9527** [comment]
But I don't understand why `Updating a new graph in one request` has so many complicated operation

**Comment by Doom9527** [comment]
I use many-to-many association tables to design the JPA model:
```
@Entity
@Table(name = "Node")
@Include(rootLevel = true, name = "node", description = "node entity", friendlyName = "node")
public class Node {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    private String name;

    @ManyToMany(mappedBy = "nodes")
    private Set<Link> links = new HashSet<>();
}
```
```
@Entity
@Table(name = "Link")
@Include(rootLevel = true, name = "link", description = "graph link", friendlyName = "link")
public class Link {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    private String label;

    @ManyToMany
    @JoinTable(
            name = "NodeLink",
            joinColumns = @JoinColumn(name = "link_id"),
            inverseJoinColumns = @JoinColumn(name = "node_id")
    )
    private Set<Node> nodes = new HashSet<>();
}

```

The GraphQL query will like:
Creating a new graph in one request
```
mutation createGraph {
    link(
        op: UPSERT
        data: { label: "label", nodes: [{ name: "NodeA" }, { name: "NodeB" }] }
    ) {
        edges {
            node {
                id
                label
                nodes {
                    edges {
                        node {
                            id
                            name
                        }
                    }
                }
            }
        }
    }
}
```

**Comment by Doom9527** [comment]
I set the `rootLevel` for both `Node` and `Link` to true, and if I set the `rootLevel` for the former to false, I will consider creating a single `Node` by adding a sentinel to the latter

**Comment by Doom9527** [comment]
I set the `rootLevel` for both `Node` and `Link` to true, and if I set the `rootLevel` for the former to false, I will consider adding a sentinel to the latter to create a single `Node`

**Comment by QubitPi** [comment]
Looks better.

- Semantically, node and link cannot be many-to-many. __A link must have exactly one source node and exactly one target node (although source and target can be the same)__. Take a look at the example below:

  ![two-nodes](https://github.com/paion-data/elide/assets/16126939/1135a3bd-efeb-4bf0-bfba-9dc0ab0641e2)

- Nexus Graph needs to be able to fetch all nodes and links of a graph. In the new data model, how would we do that? Specifically, a user can have many knowledge graphs. Each graph has __n__ nodes, __m__ links, and a graph ID

- [`joinColumns` is not portable across database systems](https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2#a14922). This would become a big trouble for API libraries like Elide:

  ![page](https://github.com/paion-data/elide/assets/16126939/8438e588-ba01-4b5e-af38-db9c8b61b31f)

- Another issue I saw with the query creating a new graph is: how would we reference newly created nodes ID's in one mutation? Because we would need them to determine the source and target nodes of a link. For example, if we need a unidirectional link between `NodeA` and `NodeB`, how to create a link like this?
